File: rent_assistant/apps/crawler/lib/crawler/sources/olx.ex
defmodule Crawler.Sources.Olx do
  defmodule Parser do
    @behaviour Crawler.Parser

    require Logger

    def parse_index_listings_page(index_page_html) do
      {:ok, document} = Floki.parse_document(index_page_html)

      Floki.find(document, "div[data-cy=\"l-card\"] a")
      |> Floki.attribute("href")
      |> Enum.filter(fn link -> String.starts_with?(link, "/d/") end)
      |> Enum.map(fn link -> Enum.join(["https://www.olx.pl", link]) end)
      |> Enum.reverse()
    end

    def parse_listing_page(listing_page_html) do
      {:ok, document} = Floki.parse_document(listing_page_html)

      init_script_text =
        Floki.find(document, "script#olx-init-config")
        |> Floki.text(js: true)

      match_data =
        Regex.named_captures(~r/__PRERENDERED_STATE__= "(?<init_json>.*)";/, init_script_text)

      init_json =
        match_data["init_json"]
        |> String.replace(~r/\\\\/, "\\")
        |> String.replace(~r/\\\"/, "\"")
        |> String.replace(~r/\\\\u002F/, "/")
        |> String.replace(~r/\\\\u003C/, "<")
        |> String.replace(~r/\\\\u003E/, ">")
        |> String.replace(~r/\\\\u0026/, "&")
        |> String.replace(~r/\\\\u0027/, "'")
        |> String.replace(~r/\\\\u0022/, "\"")
        |> String.replace(~r/\\\\\"/, "\"")
        |> String.replace(~r/\\\\\n/, "\n")

      case Poison.decode(init_json) do
        {:ok, json} ->
          ad_data = json["ad"]["ad"]

          if ad_data["id"] do
            %{
              url: ad_data["url"],
              source: "olx",
              address: ad_data["location"]["pathName"],
              price: ad_data["price"]["regularPrice"]["value"],
              area: parse_area(ad_data),
              source_id: "#{ad_data["id"]}",
              rooms: parse_rooms(ad_data),
              source_created_at: ad_data["createdTime"] |> parse_date,
              source_updated_at: ad_data["lastRefreshTime"] |> parse_date,
              location: parse_coordinates(ad_data),
              images: ad_data["photos"],
              description: Floki.text(ad_data["description"])
            }
          else
            Logger.warn("Unable to find listing id in json #{inspect(init_json)}")
            nil
          end

        {:error, error} ->
          Logger.warn("Unable to decode listing page: #{inspect(error)}")
          nil
      end
    end

    defp parse_area(ad_data) do
      case ad_data["params"] do
        nil ->
          Logger.warn("Unable to parse area: #{inspect(ad_data)}")

        params ->
          Enum.find_value(params, fn param ->
            if param["key"] == "m" do
              param["normalizedValue"] |> Integer.parse() |> elem(0)
            end
          end)
      end
    end

    defp parse_rooms(ad_data) do
      case ad_data["params"] do
        nil ->
          Logger.warn("Unable to parse rooms: #{inspect(ad_data)}")

        params ->
          Enum.find_value(params, fn param ->
            if param["key"] == "rooms" do
              if param["value"] == "Kawalerka" do
                1
              else
                param["value"] |> String.split() |> List.first() |> String.to_integer()
              end
            end
          end)
      end
    end

    defp parse_date(date) do
      {:ok, date, _} = DateTime.from_iso8601(date)
      date
    end

    defp parse_coordinates(ad_data) do
      [ad_data["map"]["lat"], ad_data["map"]["lon"]]
    end
  end
end

File: apps/crawler/app/parser/olx.rb
require 'parser'

class Olx < Parser
  def parse_index
    get_html(@url).css("div[data-cy=\"l-card\"] a[href^='/d/']").each_with_object([]) do |offer, results|
      next if offer.css("div[data-testid='adCard-featured']").any?
      href = offer.attr('href')
      results << { url: "https://www.olx.pl#{href}" }
    end.reverse
  end

  def parse_listing(index_listing)
    html = get_html(index_listing[:url])
    init_script_text = html.at("script#olx-init-config").text

    init_json_match = init_script_text.match(/__PRERENDERED_STATE__= "(?<init_json>.*)";/)
    init_json = init_json_match["init_json"]
      .gsub(/\\\\/, "\\")
      .gsub(/\\\"/, "\"")
      .gsub(/\\\\u002F/, "/")
      .gsub(/\\\\u003C/, "<")
      .gsub(/\\\\u003E/, ">")
      .gsub(/\\\\u0026/, "&")
      .gsub(/\\\\u0027/, "'")
      .gsub(/\\\\u0022/, "\"")
      .gsub(/\\\\\"/, "\"")
      .gsub(/\\\\\n/, "\n")

    begin
      json = JSON.parse(init_json)
    rescue JSON::ParserError => e
      $logger.warn("Unable to decode listing page: #{e.message}")
      return nil
    end

    ad_data = json.dig("ad", "ad")
    unless ad_data
      Rollbar.warn("Unable to find ad data in json #{init_json}")
      return nil
    end
    
    if ad_data["id"]
      {
        url: ad_data["url"],
        city: ad_data.dig("location", "cityNormalizedName"),
        address: ad_data.dig("location", "pathName"),
        price: ad_data.dig("price", "regularPrice", "value"),
        additional_price: dig_additional_price(ad_data),
        area: dig_area(ad_data) || 0,
        rooms: dig_rooms(ad_data) || 0,
        location: dig_location(ad_data),
        images: ad_data["photos"],
        description: ad_data.dig("description"),
        currency: ad_data.dig("price", "regularPrice", "currencyCode"),
        source:{
          id: ad_data["id"],
          created_at: ad_data["createdTime"],
          updated_at: ad_data["lastRefreshTime"],
        }
      }
    else
      $logger.warn("Unable to find listing id in json #{init_json}")
      nil
    end
  end

  def dig_area(ad_data)
    return nil if ad_data.nil? || ad_data["params"].nil? || ad_data["params"].empty?
  
    param = ad_data["params"].find { |param| param["key"] == "m" }
    
    return nil if param.nil? || param["normalizedValue"].nil?
  
    param["normalizedValue"].to_i
  end

  def dig_rooms(ad_data)
    rooms_val = ad_data.dig("params").find { |param| param["key"] == "rooms" }
    return nil unless rooms_val

    rooms_eng = rooms_val["normalizedValue"]
    case rooms_eng
    when "one"
      1
    when "two"
      2
    when "three"
      3
    when "four"
      4
    when "five"
      5
    else
      nil
    end
  end

  def dig_location(ad_data)
    map = ad_data['map']
    [map['lat'], map['lon']] if map
  end

  def dig_additional_price(ad_data)
    rent_param = ad_data['params'].find { |param| param["key"] == "rent" }
    return 0 unless rent_param

    rent_param["normalizedValue"].to_i
  end
end
